gradle-baseline-java/src/main/groovy/com/palantir/baseline/tasks/CheckExplicitSourceCompatibilityTask.java

```diff
@@ -0,0 +1,110 @@
+import org.gradle.api.plugins.JavaPluginConvention;
+import org.gradle.api.provider.Property;
+import org.gradle.api.publish.PublishingExtension;
+import org.gradle.api.specs.Spec;
+import org.gradle.api.tasks.TaskAction;
+import org.gradle.api.tasks.options.Option;
+
+/**
+ * By default, Gradle will infer sourceCompat based on whatever JVM is currently being used to evaluate the
+ * build.gradle files. This is bad for reproducibility because if we make an automated PR to upgrade the Java major
+ * version (e.g. 11 -> 15) then a library might unintentionally start publishing jars containing Java15 bytecode!
+ *
+ * Better to just require everyone to specify sourceCompatibility explicitly!
+ */
+public class CheckExplicitSourceCompatibilityTask extends DefaultTask {
+
+    private final Property<Boolean> shouldFix;
+
+    @Inject
+    public CheckExplicitSourceCompatibilityTask(ObjectFactory objectFactory) {
+        setGroup("Verification");
+        setDescription("Ensures build.gradle specifies sourceCompatibility explicitly, otherwise it is inferred based"
+                + " on $JAVA_HOME which is fragile.");
+        this.shouldFix = objectFactory.property(Boolean.class);
+        this.shouldFix.set(false);
+
+        onlyIf(new Spec<Task>() {
+            @Override
+            public boolean isSatisfiedBy(Task element) {
+                // sometimes people apply the 'java' plugin to projects that doesn't actually have any java code in it
+                // (e.g. the root project), so if they're not publishing anything, then we don't bother enforcing the
+                // sourceCompat thing
+                PublishingExtension publishing = getProject().getExtensions().findByType(PublishingExtension.class);
+                return publishing != null;
+            }
+        });
+    }
+
+    @Option(option = "fix", description = "Whether to apply the suggested fix to build.gradle")
+    public final void setShouldFix(boolean value) {
+        shouldFix.set(value);
+    }
+
+    @TaskAction
+    public final void taskAction() throws IOException {
+        // We're doing this naughty casting because we need access to the `getRawSourceCompatibility` method.
+        org.gradle.api.plugins.internal.DefaultJavaPluginConvention convention =
+                (org.gradle.api.plugins.internal.DefaultJavaPluginConvention)
+                        getProject().getConvention().getPlugin(JavaPluginConvention.class);
+
+        if (convention.getRawSourceCompatibility() != null) {
+            // In theory, users could configure the fancy new 'java toolchain' as an alternative to explicit
+            // sourceCompatibility, but there's no method to access this yet (as of Gradle 6.8).
+            return;
+        }
+
+        if (shouldFix.get()) {
+            Files.write(
+                    getProject().getBuildFile().toPath(),
+                    Collections.singletonList(String.format("%nsourceCompatibility = %s", JavaVersion.current())),
```

Comment:
Should this also add a trailing newline at the end of the file?

```suggestion
                    Collections.singletonList(String.format("%nsourceCompatibility = %s%n", JavaVersion.current())),
```